util/matchers.go

```diff
@@ -10,7 +10,8 @@ import (
 // SplitFiltersAndMatchers splits empty matchers off, which are treated as filters, see #220
 func SplitFiltersAndMatchers(allMatchers []*metric.LabelMatcher) (filters, matchers []*metric.LabelMatcher) {
 	for _, matcher := range allMatchers {
-		if matcher.Match("") {
+		// Test for empty filters. Other metrics such as NotEqual are not filters.
```

Comment:
As mentioned in the original comment: "SplitFiltersAndMatchers splits empty matchers off, which are treated as filters, see #220"

Unfortunately `{foo!="bar"}` matches `""`, which means it also gets split off as a filter and we don't get a chance to apply optimisations before we return any chunks. I came across this with queries such as `{__name__!="foo"}` and with our optimisation before fetching chunks not being made because the matcher is filtered out from matchers.

Is this correct? I can add some more comments explaining if it is.